codegen/src/function.rs

```diff
@@ -611,15 +655,24 @@ impl ExportedFn {
                             ..
                         }) => match elem.as_ref() {
                             &syn::Type::Path(ref p) if p.path == str_type_path => {
-                                is_str_ref = true;
+                                is_string = true;
+                                is_ref = true;
                                 quote_spanned!(arg_type.span()=>
                                                    mem::take(args[#i])
                                                    .clone().cast::<ImmutableString>())
                             }
                             _ => panic!("internal error: why wasn't this found earlier!?"),
                         },
+                        &syn::Type::Path(ref p) if p.path == string_type_path => {
+                            is_string = true;
+                            is_ref = false;
+                            quote_spanned!(arg_type.span()=>
+                                                   mem::take(args[#i])
+                                                   .clone().cast::<#arg_type>())
```

Comment:
You appear to have created a `Dynamic::take_immutable_string` method for just this purpose. Would using that be cheaper than a `clone()`?

Reply:
Yeah, good idea!